EssentialApp/EssentialAppTests/ImageCommentsUIIntegrationTests.swift

```
@@ -0,0 +1,330 @@
+		let (sut, loader) = makeSUT()
+
+		let comment0 = makeComment(
+			message: "a message",
+			username: "a username"
+		)
+
+		let comment1 = makeComment(
+			message: "another message",
+			username: "another username"
+		)
+
+		sut.loadViewIfNeeded()
+		assertThat(sut, isRendering: [ImageComment]())
+
+		let comments0 = [comment0]
+		loader.completeLoading(with: comments0, at: 0)
+		assertThat(sut, isRendering: comments0)
+
+		sut.simulateUserInitiatedReload()
+		let comments1 = [comment0, comment1]
+		loader.completeLoading(with: comments1, at: 1)
+		assertThat(sut, isRendering: comments1)
+	}
+
+	func test_loadCompletion_rendersSuccessfullyLoadedEmptyCommentsAfterNonEmptyComments() {
+		let (sut, loader) = makeSUT()
+
+		let comments = [makeComment()]
+
+		sut.loadViewIfNeeded()
+		loader.completeLoading(with: comments)
+		assertThat(sut, isRendering: comments)
+	}
+
+	func test_loadCompletion_doesNotAlterCurrentRenderingStateOnError() {
+		let (sut, loader) = makeSUT()
+		let comments = [makeComment()]
+
+		sut.loadViewIfNeeded()
+		loader.completeLoading(with: comments, at: 0)
+		assertThat(sut, isRendering: comments)
+
+		sut.simulateUserInitiatedReload()
+		loader.completeLoadingWithError(at: 1)
+		assertThat(sut, isRendering: comments)
+	}
+
+	func test_loadCompletion_dispatchesFromBackgroundToMainThread() {
+		let (sut, loader) = makeSUT()
+		sut.loadViewIfNeeded()
+
+		let exp = expectation(description: "Wait for background queue")
+		DispatchQueue.global().async {
+			loader.completeLoading(at: 0)
+			exp.fulfill()
+		}
+
+		wait(for: [exp], timeout: 1.0)
+	}
+
+	func test_loadCompletion_rendersErrorMessageOnErrorUntilNextReload() {
+		let (sut, loader) = makeSUT()
+
+		sut.loadViewIfNeeded()
+		XCTAssertEqual(sut.errorMessage, nil)
+
+		loader.completeLoadingWithError(at: 0)
+		XCTAssertEqual(sut.errorMessage, ImageCommentsPresenter.errorMessage)
+
+		sut.simulateUserInitiatedReload()
+		XCTAssertEqual(sut.errorMessage, nil)
+	}
```

It's missing a test/assertion checking that `tapping the error view will hide the error`.

---

Addressed in https://github.com/essentialdevelopercom/ios-lead-essentials-image-comments-challenge/pull/17/commits/8dddb152aa4e30921d9407872f8de89c11f03443.